File: ts/shared/impl/KeyEvent.ts
import type { I_KeyEvent } from "@shared/interfaces/I_KeyEvent"
import { KeyState } from "@shared/enum/KeyState"
import { Key } from "./Key"
export class KeyEvent implements I_KeyEvent {
	constructor(){}
	protected __init__(...args: Parameters<typeof KeyEvent.new>){
		const z = this
		z.key = args[0]
		z.state = args[1]
		return z
	}

	static new(key:Key, state:KeyState){
		const z = new this()
		z.__init__(key, state)
		return z
	}

	key: Key
	state: KeyState
}

File: cc/src/impl/KeyEvent.cc
#include "./KeyEvent.h"
#include "enum/KeyState.h"
namespace _ {

bool KeyEvent::isKeyDown(I_KeyEvent& e, I_Key& key){
	// println("e.key_()->code_(): ",e.key_()->code_());
	// println("key.code_(): ",key.code_());
	// println("e.state_() == KeyState::down: ",e.state_() == KeyState::down);
	// println("e.state_() == KeyState::up: ",e.state_() == KeyState::up);
	if(
		e.key_()->code_() == key.code_()
		&& e.state_() == KeyState::down
	){
		return true;
	}
	return false;
}

bool KeyEvent::isKeyUp(I_KeyEvent& e, I_Key& key){
	// println("e.key_()->code_(): ",e.key_()->code_());
	// println("key.code_(): ",key.code_());
	// println("e.state_() == KeyState::down: ",e.state_() == KeyState::down);
	// println("e.state_() == KeyState::up: ",e.state_() == KeyState::up);
	if(
		e.key_()->code_() == key.code_()
		&& e.state_() == KeyState::up
	){
		return true;
	}
	return false;
}
bool KeyEvent::eq(I_KeyEvent &e1, I_KeyEvent &e2){
	return e1.key_()->code_() == e2.key_()->code_() && e1.state_() == e2.state_();
}


}//namespace _